### ISSUE TITLE
how to distinguish MinGW, Cygwin GCC and MSYS GCC builds?

### ISSUE THREAD

**Issue by SSE4** [conan-io/conan#2120]
most of recipes are now using the following statement to identify GCC:

```
if self.settings.os == "Windows" and self.settings.compiler == "gcc":
    <MinGW build here>
```

however, there are actually multiple cases:

1) native Windows binaries built with MinGW, depends on MinGW runtime (like libgcc.dll, etc)
2) binaries for Cygwin environment, or at least dependent on cygwin.dll
2) binaries for MSYS environment, or at least dependent on msys.dll

what combination of settings allow to distinguish them? something like compiler.runtime or may be their own OS values?

/cc @dimi309 @solvingj

**Comment by lasote** [comment]
Could it be a `runtime` subsetting of the `gcc` compiler?

**Comment by SSE4** [comment]
it could be, unless it conflicts/intersects with libstdc++/libstdc++11
is there some standardised practice already in place?

**Comment by lasote** [comment]
What you mean?

**Comment by SSE4** [comment]
I mean, package authors might be already using some practice to distinguish MinGW, Cygwin's GCC and MSYS's GCC

**Comment by lasote** [comment]
I don't think so, at least not a settings level.

**Comment by lasote** [comment]
Mhhh, probably should be a subsetting of Windows. You could use the Visual Studio toolchain with the build tools from msys for example. Many things to investigate...

**Comment by SSE4** [comment]
targeting MSYS environment explicitly and using tools from MSYS (e.g. make, autotools) to build for Visual Studio or MinGW are different things for sure.

**Comment by dimi309** [comment]
I'm installing cygwin to experience the issues first-hand. I'm more of a mingw person myself :) One thought I've had, without meaning to discourage any improvements, is that perhaps things are ok the way they are? In all cases it is true that the compiler is gcc or clang and the build is running on a specific operating system (Windows). So conan accurately relfects this situation with its package settings. The mixup between these builds comes from the fact that they are not adequately emulating how they would proceed in one of gcc / clang 's "native" environments and their produced binaries have certain particularities. I wonder if this is a conan issue, or an issue with cygwin and mingw themselves. If the latter is the case, conan will hurt itself if it tries to fix things by complicating its own logic with a sort of patch for these minority-used build systems. I say this without contempt as I am a fan of mingw myself and always give it a tryat least to make my packages work with mingw before publishing them..

**Comment by memsharded** [comment]
I also think they are subsettings of the Windows ``settings.os``, because they are different subsystems, and binaries are indeed different and incompatible, in most cases. For me:

```yaml
os: Windows: 
         subsystem: [None, cygwin, WLS, msys, mingw]
```

Then, as there are some settings that might cause binaries to be fully compatible, it could be useful to have some ``package_id()`` default behavior that would implement the logic (like msys + CL compiler => None)

**Comment by dimi309** [comment]
That might actually work..

**Comment by lasote** [comment]
Not sure about it. The real subsystems are Msys2 and Cygwin. You can build applications for these subsystems that will only run in these subsystems (or at least with their own DLL dependencies, haven't enough experience) but building a MSVC application with Cygwin or Msys is equivalent to use a different "build system". So, probably shouldn't be modeled as setting. In that case, the subsystem should be None. 

With msys and mingw it is similar. You can build native Windows applications starting msys in "mingw" mode, so you are not building msys2 apps. 

The original question by @sse4 is pretty accurate, but I would model it as a subsystem of Windows, irrespective of if it's build with gcc, clang or other.

So you have the compiler setting (gcc) and optional in windows a subsetting for the subsystem, msys and Cygwin. 

About the WLS I would need more info.

**Comment by dimi309** [comment]
Thanks for that explanation. A lot of things about msys and cygwin are clearer to me now. So, I suppose that, mingw-make builds (without msys) would also have 'None' as subsystem because they are native to windows, correct? And, knowing that WLS is a linux kernel running under windows and being interpreted to the windows kernel, wouldn't that mean that we should just consider it a different linux distribution and not a windows subsystem?

**Comment by lasote** [comment]
To the first question, correct.
To the second one, probably yes, I haven't try it but probably it has nothing to do with Windows, its probably just a VM or something similar. I will search for more information.

**Comment by dimi309** [comment]
Ok, thanks a lot for taking the time to share your thoughts with us on this.

**Comment by lasote** [comment]
About the `mingw-make` you mentioned. You could even use the `make` and other tools from MSYS or CYGWIN, but building a native Windows application, the `mingw-make` is a native Windows `make` application but it doesn't work in all cases. The `make` from MSYS  and CYGWIN are not native applications but can build native applications.
That's why you can safely use the classic gnu tools from these subsystems but build an application that just depend on the pure native windows runtime libraries.

However, for the conan recipes there is an issue to solve, when you are going to use for instance, MSYS2 to build a library with make, you need to prepare some path in the style of MSYS2, the same for cygwin. We have the cool `tools.unix_path` with a parameter to specify different flavors, great contribution of @SSE4, but you need to know the subsystem you are going to use first.

The Bincrafters approach, as I could have seen so far, is to model the subsystem as an option, and then use the `package_id` to not generate different packages. Using also a build_require in the recipe to retrieve the MSYS or CYGWIN.

I think the correct approach would be:

- Do not use `options` for that.
- Do not use `build_requires` in the recipes for that, but be prepared to use them as a profile build_require.
- Just invoke the `make` you have available through a `bash`. And here is the `issue`, you need to prepare some path, for the subsystem.  I think we could try to detect the subsystem with some Conan helper invoking the `bash` first, to know the subsystem and adjust the needed things.

I will keep trying these things and will report any progress.

**Comment by dimi309** [comment]
@SSE4 @solvingj @sztomi This is all very interesting insights for our discussion (https://github.com/bincrafters/community/issues/13) Just pinging you to let you know. I am keeping notes as I am still working on that summary I want to prepare about the mingw/msys/cygwin/wsl situation.

**Comment by solvingj** [comment]
Glad to see this might get a first class feature. Our option and build requires approach came before we were familiar with profiles. I can see a rabbithole of features that could be added around the Subsystems, but I think just detecting and being able to pass commands appropriately and generically is the most important.

**Comment by sigmoidal** [comment]
don't forget that the user env path may be containing a mixture of these tools, or partial installs. maybe let the the user provide not only the subsystem, but also its location, then setup a pristinr env from scratch.

**Comment by lasote** [comment]
I the user have a mess of environment we won't be able to cover all the cases, for me, the goal is to tell the user, ok, apply this `build_require` and you will be good. About the location, probably the current `CONAN_BASH_PATH` as an entry point sounds good, let's see if we need more things.

**Comment by sigmoidal** [comment]
Some users install things they don't even know that they come with gnu tools and/or toolchains (see Strawberry Perl). It is not a mess because the user has been messy, it is because some applications assume it is ok to take a bunch of binaries they need drop them into an installer and then slap it into the user's PATH (which most users may not even notice).

I see what you are thinking, ok. I agree it is not possible to cover all cases, but lets agree that the build_require can't guarantee the user *will* be good either, lets assume he *may* be good, i.e. let a user have a way out if what you are providing is not working for him and/or would like to take control.

The idea of using CONAN_BASH_PATH for providing the hints necessary for detecting the environment sounds awesome.

thank you.

**Comment by solvingj** [comment]
I would point out that the definition of "Conan User" here is important. It is distinct from "Author of a `build_requires` package".

**Comment by memsharded** [comment]
This has been released in the 1.0 beta: http://docs.conan.io/en/latest/conan1.0.html 
- There are MSYS, Cygwin, WLS, Mingw WIndows subsystems as **sub-settings** of the Windows OS setting.

Still needs some docs, but feedback is very welcome!
